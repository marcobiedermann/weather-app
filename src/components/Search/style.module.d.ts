declare const styles: {
  readonly search__field: string;
  readonly search__input: string;
  readonly search__submit: string;
};
export = styles;
