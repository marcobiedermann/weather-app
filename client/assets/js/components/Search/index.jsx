import { Field, Form, withFormik } from 'formik';
import PropTypes from 'prop-types';
import React from 'react';

const Search = (props) => {
  const { isSubmitting } = props;

  return (
    <Form>
      <div>
        <Field name="query" />
      </div>
      <div>
        <button type="submit" disabled={isSubmitting}>
          Search
        </button>
      </div>
    </Form>
  );
};

Search.propTypes = {
  isSubmitting: PropTypes.bool,
};

Search.defaultProps = {
  isSubmitting: false,
};

export default withFormik({
  handleSubmit: (values, { props, setSubmitting }) => {
    props.fetchSearch(values.query);
    setSubmitting(false);
  },
  mapPropsToValues: () => ({
    query: '',
  }),
})(Search);
