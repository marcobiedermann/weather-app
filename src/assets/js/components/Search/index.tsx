import classNames from 'classnames';
import { Field, Form, Formik, FormikHelpers } from 'formik';
import React, { FC } from 'react';
import { useTranslation } from 'react-i18next';
import styles from '../Settings/style.module.css';

interface Values {
  query: string;
}

export interface SearchProps {
  initialValues: Values;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onSubmit: (values: Values, formikHelpers: FormikHelpers<Values>) => void | Promise<any>;
}

const Search: FC<SearchProps> = (props) => {
  const { t } = useTranslation();

  return (
    <Formik {...props}>
      {({ isSubmitting }) => (
        <Form>
          <div className={styles.form__field}>
            <Field name="query" className={styles.form__input} />
          </div>

          <div className={styles.form__field}>
            <button
              type="submit"
              disabled={isSubmitting}
              className={classNames(styles.form__input, styles['form__input--submit'])}
            >
              {t('translation:search')}
            </button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default Search;
