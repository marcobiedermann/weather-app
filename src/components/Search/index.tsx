import { useForm } from 'react-hook-form';
import { Search as SearchIcon } from 'react-feather';
import { useTranslation } from 'react-i18next';
import styles from './style.module.css';

interface FormData {
  query: string;
}

interface SearchProps {
  defaultValues?: FormData;
  onSubmit: (data: FormData) => void;
}

function Search(props: SearchProps): JSX.Element {
  const { defaultValues, onSubmit } = props;
  const { handleSubmit, register } = useForm<FormData>({
    defaultValues,
  });
  const { t } = useTranslation();

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={styles.search}>
      <input
        {...register('query')}
        className={styles.search__input}
        placeholder="Search location here"
      />

      <button type="submit" className={styles.search__submit}>
        <span style={{ display: 'none' }}>{t('search')}</span>
        <SearchIcon />
      </button>
    </form>
  );
}

export default Search;
