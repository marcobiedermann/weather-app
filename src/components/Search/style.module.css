.search {
  align-items: center;
  background-color: #fcfcfc;
  border-radius: 0.25em;
  color: #8f9298;
  display: flex;
}

.search__input {
  flex-grow: 1;
  padding: 1em 1.5em;
}

.search__submit {
  flex-grow: 0;
  order: -1;
  padding: 1em;
}
