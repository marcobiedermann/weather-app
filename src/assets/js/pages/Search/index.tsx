import { FormikHelpers } from 'formik';
import React, { FC, useState } from 'react';
import { Helmet } from 'react-helmet';
import useSWR from 'swr';
import Loader from '../../components/Loader';
import Results from '../../components/Results';
import Search from '../../components/Search';

interface SearchPageQuery {
  message: string;
  cod: string;
  count: number;
  list: List[];
}

interface List {
  id: number;
  name: string;
  coord: Coord;
  main: Main;
  dt: number;
  wind: Wind;
  sys: Sys;
  rain: null;
  snow: null;
  clouds: Clouds;
  weather: Weather[];
}

interface Clouds {
  all: number;
}

interface Coord {
  lat: number;
  lon: number;
}

interface Main {
  temp: number;
  feels_like: number;
  temp_min: number;
  temp_max: number;
  pressure: number;
  humidity: number;
}

interface Sys {
  country: string;
}

interface Weather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

interface Wind {
  speed: number;
  deg: number;
}

interface Values {
  query: string;
}

const SearchPage: FC = () => {
  const [query, setQuery] = useState('London, GB');
  const { data, error } = useSWR<SearchPageQuery>(`/find?q=${query}`);

  if (error) {
    return <div>Error: {error}</div>;
  }

  const initialValues: Values = {
    query,
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function onSubmit(values: Values, formikHelpers: FormikHelpers<Values>): void | Promise<any> {
    const { setSubmitting } = formikHelpers;

    setQuery(values.query);
    setSubmitting(false);
  }

  return (
    <div>
      <Helmet>
        <title>Search</title>
      </Helmet>
      <Search initialValues={initialValues} onSubmit={onSubmit} />
      {data ? <Results results={data.list} /> : <Loader />}
    </div>
  );
};

export default SearchPage;
