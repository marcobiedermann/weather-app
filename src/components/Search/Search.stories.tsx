import type { Meta, StoryObj } from '@storybook/react';
import Search from '.';

const meta: Meta<typeof Search> = {
  component: Search,
  args: {
    onSubmit: (data) => console.log(data),
  },
};

type Story = StoryObj<typeof Search>;

const Default: Story = {};

export { Default };
export default meta;
