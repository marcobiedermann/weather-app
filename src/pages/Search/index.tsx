import React, { useState } from 'react';
import { Helmet } from 'react-helmet';
import Error from '../../components/Error';
import Loader from '../../components/Loader';
import Results from '../../components/Results';
import Search, { FormData } from '../../components/Search';
import { useFind } from '../../hooks';

function SearchPage(): JSX.Element {
  const [query, setQuery] = useState('London, GB');
  const { data, isError, error } = useFind(query);

  if (isError) {
    return <Error message={error.message} />;
  }

  const defaultValues: FormData = {
    query,
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function onSubmit(data: FormData): void | Promise<any> {
    const { query } = data;

    setQuery(query);
  }

  return (
    <>
      <Helmet>
        <title>Search</title>
      </Helmet>
      <Search defaultValues={defaultValues} onSubmit={onSubmit} />
      {data ? <Results results={data.list} /> : <Loader />}
    </>
  );
}

export default SearchPage;
