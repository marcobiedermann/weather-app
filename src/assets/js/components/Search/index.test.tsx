import React from 'react';
import renderer from 'react-test-renderer';
import Search from '.';

describe('components/Search', () => {
  it('renders correctly', () => {
    expect.assertions(1);

    const tree = renderer
      .create(<Search initialValues={{ query: '' }} onSubmit={jest.fn()} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
