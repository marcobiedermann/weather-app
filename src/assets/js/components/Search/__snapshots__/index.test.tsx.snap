// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Search renders correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div>
    <input
      name="query"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </div>
  <div>
    <button
      disabled={false}
      type="submit"
    >
      translation:search
    </button>
  </div>
</form>
`;
